onnx/backend/test/cpp/driver/test_driver.cc

```diff
@@ -0,0 +1,180 @@
+int TestDriver::fetchSingleTestCase(const std::string& case_dir) {
+  std::string model_name = case_dir;
+  model_name += "model.onnx";
+  if (fileExist(model_name)) {
+    TestCase test_case;
+    test_case.model_filename_ = model_name;
+    test_case.model_dirname_ = case_dir;
+    int case_count = 0;
+    for (;; case_count++) {
+      std::vector<std::string> input_filenames, output_filenames;
+      std::string input_name, output_name;
+      std::string case_dirname = case_dir;
+      case_dirname += "test_data_set_" + std::to_string(case_count);
+
+      output_name = case_dirname + "/output_" + "0" + ".pb";
+      if (!fileExist(output_name)) {
+        break;
+      }
+
+      for (int data_count = 0;; data_count++) {
+        input_name =
+            case_dirname + "/input_" + std::to_string(data_count) + ".pb";
+        output_name =
+            case_dirname + "/output_" + std::to_string(data_count) + ".pb";
+        if (!fileExist(input_name) && !fileExist(input_name)) {
+          break;
+        }
+        if (fileExist(input_name)) {
+          input_filenames.push_back(input_name);
+        }
+        if (fileExist(output_name)) {
+          output_filenames.push_back(output_name);
+        }
+      }
+      TestData test_data(input_filenames, output_filenames);
+      test_case.test_data_.push_back(test_data);
+    }
+    testcases_.push_back(test_case);
+  }
+  return 0;
+}
+
+//load all test data in target_dir to _testcases
+int TestDriver::fetchAllTestCases(const std::string& _target_dir){
+	std::string target_dir = _target_dir;
+	if (target_dir == ""){
+		target_dir = default_dir_;
+	}
+	DIR* directory = opendir(target_dir.c_str());
+	if (directory == NULL){
+		fprintf(stderr, "Error: cannot open directory %s when loading test data: %s\n",
+			target_dir.c_str(), strerror(errno));
+		return -1;
+	}
+
+	while (true){
+		errno = ERRNO_DIR_END;
+		struct dirent* entry = readdir(directory);
+		if (entry == NULL){
+			if (errno != ERRNO_DIR_END){
+				fprintf(stderr, "Error: cannot read directory %s when loading test data: %s\n",
+					target_dir.c_str(), strerror(errno));
+				return -1;
+			} else{
+				break;
+			}
+		}
+                std::string entry_dname = entry->d_name;
+                entry_dname = _target_dir + '/' + entry_dname + "/";
+                fetchSingleTestCase(entry_dname);
+        }
+cleanup:
+	if (directory != NULL){
+		if (closedir(directory) != 0){
+			fprintf(stderr, "Warning: failed to close directory %s when loading test data: %s\n",
+				target_dir.c_str(), strerror(errno));
+			return -1;
+		}
+	}
+	return 0;
+}
+
+std::vector<TestCase> getTestCase(const std::string& location){
+	TestDriver test_driver;
+	test_driver.fetchAllTestCases(location);
+	return test_driver.testcases_;
+}
+
+void loadSingleFile(const std::string& filename, std::string& filedata){
+	std::ifstream f(filename, std::ifstream::in);
+	if (f.is_open()){
+			std::string line;
+		getline(f, filedata);
```

Comment:
Do you use this function only for text files?

Reply:
This is actually a protobug file, but I would like to read it byte by byte

Reply:
There seems to be one-liner: https://stackoverflow.com/questions/3203452/how-to-read-entire-stream-into-a-stdstring

Reply:
Don't read binary files line-by-line! I bet you don't want to debug what happens when some lines end with `\n`, others with `\r\n`.